File: frameworks/play-graphql/test/TestFactory.scala
package test

import java.time.Instant
import java.util.UUID
import scala.collection.mutable

/**
 * Test data models
 */
case class TestUser(
  id: String,
  pkUser: Int,
  username: String,
  fullName: String,
  bio: Option[String],
  createdAt: Instant,
  updatedAt: Instant
)

case class TestPost(
  id: String,
  pkPost: Int,
  fkAuthor: Int,
  title: String,
  content: String,
  createdAt: Instant,
  updatedAt: Instant,
  author: Option[TestUser]
)

case class TestComment(
  id: String,
  pkComment: Int,
  fkPost: Int,
  fkAuthor: Int,
  content: String,
  createdAt: Instant,
  author: Option[TestUser],
  post: Option[TestPost]
)

/**
 * In-memory test factory for isolated tests
 */
class TestFactory {
  private val users = mutable.Map[String, TestUser]()
  private val posts = mutable.Map[String, TestPost]()
  private val comments = mutable.Map[String, TestComment]()

  private var userCounter = 0
  private var postCounter = 0
  private var commentCounter = 0

  def createUser(
    username: String,
    email: String,
    fullName: String,
    bio: Option[String] = None
  ): TestUser = {
    userCounter += 1
    val user = TestUser(
      id = UUID.randomUUID().toString,
      pkUser = userCounter,
      username = username,
      fullName = fullName,
      bio = bio,
      createdAt = Instant.now(),
      updatedAt = Instant.now()
    )
    users(user.id) = user
    user
  }

  def createPost(
    authorId: String,
    title: String,
    content: String = "Default content"
  ): TestPost = {
    val author = users.getOrElse(authorId,
      throw new RuntimeException(s"Author not found: $authorId"))

    postCounter += 1
    val post = TestPost(
      id = UUID.randomUUID().toString,
      pkPost = postCounter,
      fkAuthor = author.pkUser,
      title = title,
      content = content,
      createdAt = Instant.now(),
      updatedAt = Instant.now(),
      author = Some(author)
    )
    posts(post.id) = post
    post
  }

  def createComment(
    authorId: String,
    postId: String,
    content: String
  ): TestComment = {
    val author = users.getOrElse(authorId,
      throw new RuntimeException("Author not found"))
    val post = posts.getOrElse(postId,
      throw new RuntimeException("Post not found"))

    commentCounter += 1
    val comment = TestComment(
      id = UUID.randomUUID().toString,
      pkComment = commentCounter,
      fkPost = post.pkPost,
      fkAuthor = author.pkUser,
      content = content,
      createdAt = Instant.now(),
      author = Some(author),
      post = Some(post)
    )
    comments(comment.id) = comment
    comment
  }

  def getUser(id: String): Option[TestUser] = users.get(id)

  def getPost(id: String): Option[TestPost] = posts.get(id)

  def getComment(id: String): Option[TestComment] = comments.get(id)

  def getAllUsers: Seq[TestUser] = users.values.toSeq

  def getPostsByAuthor(authorPk: Int): Seq[TestPost] =
    posts.values.filter(_.fkAuthor == authorPk).toSeq

  def getCommentsByPost(postPk: Int): Seq[TestComment] =
    comments.values.filter(_.fkPost == postPk).toSeq

  def reset(): Unit = {
    users.clear()
    posts.clear()
    comments.clear()
    userCounter = 0
    postCounter = 0
    commentCounter = 0
  }
}

object ValidationHelper {
  private val UuidRegex = "^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$".r

  def isValidUuid(value: String): Boolean =
    UuidRegex.findFirstIn(value.toLowerCase).isDefined
}

object DataGenerator {
  def generateLongString(length: Int): String = "x" * length

  def generateRandomUsername(): String = s"user_${UUID.randomUUID().toString.take(8)}"
}

File: frameworks/quarkus-graphql/src/test/java/com/velocitybench/TestFactory.java
package com.velocitybench;

import java.time.Instant;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.Collectors;

/**
 * In-memory test factory for isolated tests.
 * Thread-safe implementation using ConcurrentHashMap.
 */
public class TestFactory {

    // Data classes
    public static class TestUser {
        public final String id;
        public final int pkUser;
        public final String username;
        public final String fullName;
        public final String bio;
        public final Instant createdAt;
        public final Instant updatedAt;

        public TestUser(String id, int pkUser, String username, String fullName, String bio) {
            this.id = id;
            this.pkUser = pkUser;
            this.username = username;
            this.fullName = fullName;
            this.bio = bio;
            this.createdAt = Instant.now();
            this.updatedAt = Instant.now();
        }
    }

    public static class TestPost {
        public final String id;
        public final int pkPost;
        public final int fkAuthor;
        public final String title;
        public final String content;
        public final Instant createdAt;
        public final Instant updatedAt;
        public final TestUser author;

        public TestPost(String id, int pkPost, int fkAuthor, String title, String content, TestUser author) {
            this.id = id;
            this.pkPost = pkPost;
            this.fkAuthor = fkAuthor;
            this.title = title;
            this.content = content;
            this.author = author;
            this.createdAt = Instant.now();
            this.updatedAt = Instant.now();
        }
    }

    public static class TestComment {
        public final String id;
        public final int pkComment;
        public final int fkPost;
        public final int fkAuthor;
        public final String content;
        public final Instant createdAt;
        public final TestUser author;
        public final TestPost post;

        public TestComment(String id, int pkComment, int fkPost, int fkAuthor, String content, TestUser author, TestPost post) {
            this.id = id;
            this.pkComment = pkComment;
            this.fkPost = fkPost;
            this.fkAuthor = fkAuthor;
            this.content = content;
            this.author = author;
            this.post = post;
            this.createdAt = Instant.now();
        }
    }

    // Storage
    private final Map<String, TestUser> users = new ConcurrentHashMap<>();
    private final Map<String, TestPost> posts = new ConcurrentHashMap<>();
    private final Map<String, TestComment> comments = new ConcurrentHashMap<>();

    private final AtomicInteger userCounter = new AtomicInteger(0);
    private final AtomicInteger postCounter = new AtomicInteger(0);
    private final AtomicInteger commentCounter = new AtomicInteger(0);

    public TestUser createUser(String username, String email, String fullName, String bio) {
        int pk = userCounter.incrementAndGet();
        TestUser user = new TestUser(
            UUID.randomUUID().toString(),
            pk,
            username,
            fullName,
            bio
        );
        users.put(user.id, user);
        return user;
    }

    public TestUser createUser(String username, String email, String fullName) {
        return createUser(username, email, fullName, null);
    }

    public TestPost createPost(String authorId, String title, String content) {
        TestUser author = users.get(authorId);
        if (author == null) {
            throw new RuntimeException("Author not found: " + authorId);
        }

        int pk = postCounter.incrementAndGet();
        TestPost post = new TestPost(
            UUID.randomUUID().toString(),
            pk,
            author.pkUser,
            title,
            content,
            author
        );
        posts.put(post.id, post);
        return post;
    }

    public TestPost createPost(String authorId, String title) {
        return createPost(authorId, title, "Default content");
    }

    public TestComment createComment(String authorId, String postId, String content) {
        TestUser author = users.get(authorId);
        TestPost post = posts.get(postId);

        if (author == null) {
            throw new RuntimeException("Author not found");
        }
        if (post == null) {
            throw new RuntimeException("Post not found");
        }

        int pk = commentCounter.incrementAndGet();
        TestComment comment = new TestComment(
            UUID.randomUUID().toString(),
            pk,
            post.pkPost,
            author.pkUser,
            content,
            author,
            post
        );
        comments.put(comment.id, comment);
        return comment;
    }

    public TestUser getUser(String id) {
        return users.get(id);
    }

    public TestPost getPost(String id) {
        return posts.get(id);
    }

    public TestComment getComment(String id) {
        return comments.get(id);
    }

    public List<TestUser> getAllUsers() {
        return new ArrayList<>(users.values());
    }

    public List<TestPost> getPostsByAuthor(int authorPk) {
        return posts.values().stream()
            .filter(p -> p.fkAuthor == authorPk)
            .collect(Collectors.toList());
    }

    public List<TestComment> getCommentsByPost(int postPk) {
        return comments.values().stream()
            .filter(c -> c.fkPost == postPk)
            .collect(Collectors.toList());
    }

    public void reset() {
        users.clear();
        posts.clear();
        comments.clear();
        userCounter.set(0);
        postCounter.set(0);
        commentCounter.set(0);
    }

    // Utility classes
    public static class ValidationHelper {
        private static final String UUID_REGEX =
            "^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$";

        public static boolean isValidUuid(String value) {
            return value != null && value.toLowerCase().matches(UUID_REGEX);
        }
    }

    public static class DataGenerator {
        public static String generateLongString(int length) {
            return "x".repeat(length);
        }

        public static String generateRandomUsername() {
            return "user_" + UUID.randomUUID().toString().substring(0, 8);
        }
    }
}
